sdk/core/azure-core/inc/azure/core/paged_response.hpp

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Provides support for paged responses.
+ */
+
+#pragma once
+
+#include <cstdlib>
+#include <string>
+
+#include "azure/core/context.hpp"
+#include "azure/core/http/raw_response.hpp"
+#include "azure/core/nullable.hpp"
+
+namespace Azure { namespace Core {
+
+  /**
+   * @brief Defines the base type and behavior for a paged response.
+   *
+   * @remark The template is using for making static-inheritance.
+   *
+   * @remark T classes must implement the way to get and move to the next page.
+   *
+   * @tparam T A class type for static-inheritance.
+   */
+  template <class T> class PagedResponse {
+  private:
+    // The field used to check when the end of the response is reached. All responses from a service
+    // will always come with a payload that represents a page. The page might or might not contain
+    // elements in the page. `m_hasPage` is then turned to `false` once `MoveToNextPage` is called
+    // on the last page.
+    bool m_hasPage = true;
+
+  protected:
+    PagedResponse() = default;
+    PagedResponse(PagedResponse&&) = default;
+    PagedResponse& operator=(PagedResponse&&) = default;
+
+  public:
+    /**
+     * @brief Defines the token used to fetch the current page.
+     *
+     */
+    std::string CurrentPageToken;
```

Comment:
If `NextPageToken` is nullable, `CurrentPageToken` should also be nullable, right?

Reply:
I don't think so. CurrentPageToken is calculated based on NextPageToken.  It is not a value coming from network.

Reply:
Then what `CurrentPageToken` will be for the first page and the last page (sentinel page)?

Reply:
Not sure about the first page. Right now is empty string, so yes, looks like a sentinel.  (I need to look for Jeff´s email about this feature of the current page)
For the last page, the CurrentPage will be udpated just fine because it is set when moving to that lastPage.
For the page after the last page (the one which set the hasPage = false)  it should not matter what's at any of the Page state (content, items, current page, next page, etc), as soon as hasPage() == false, the end-user knows the page has gone after the last one and the content is useless.